packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1159 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.8
+
+import 'dart:async';
+import 'dart:collection';
+import 'dart:convert';
+
+import 'package:flutter/foundation.dart';
+
+import 'basic.dart';
+import 'binding.dart';
+import 'framework.dart';
+import 'navigator.dart';
+import 'route_notification_messages.dart';
+
+/// A piece of routing information
+///
+/// This information flows two ways, from the [RouteInformationProvider] to the
+/// [Router] or from the [Router] to [RouteInformationProvider].
+///
+/// In the former case, the [RouteInformationProvider] notifies the [Router]
+/// widget when a new [RouteInformation] is available. The [Router] widget takes
+/// these information and navigates accordingly.
+///
+/// The latter case should only happen in a web application where the [Router]
+/// reports route change back to web engine.
+class RouteInformation {
+  /// Creates a route information.
+  const RouteInformation({this.routeName, this.state});
+
+  /// Creates a route information from a serialized string.
+  factory RouteInformation.fromSerialized(String serialized) {
+    final dynamic json = jsonDecode(serialized);
```

Comment:
Why is this json encoded?

Reply:
Why does `RouteInformation` need to have an opinion at all whether and how its serializable? Isn't that up to the implementation of `RouteInformationParser` to decide how to send it over?

Reply:
If we put the serialized method to the route name provider, then we will be pushing the responsibility to figure out whether it has changed to route name provider as well because router cannot tell whether it has change by looking at the identity alone.

The RouteInformation reported from route name parser will always be different because they are usually dynamically generated.

This causes one issue, that the router also have the api to force the route name provider to report the route information back to engine even if it has not changed. we have to somehow send the reportingstatus to the route information provider, this may make the api confusing.

Reply:
we should just change System.navigation to not use JSON.